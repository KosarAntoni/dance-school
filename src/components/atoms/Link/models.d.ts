import { NavBaseProps } from 'atoms/NavBase';
import { TypographyProps } from 'atoms/Typography';

export interface LinkProps extends NavBaseProps, TypographyProps {
  color?: 'white' | 'black' | 'blue' | string;
  variant?: 'solid' | 'simple' | string;
  hasArrow?: boolean;
}
