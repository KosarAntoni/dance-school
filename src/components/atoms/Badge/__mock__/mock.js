export const BadgeDefaultClassesMock = [
  'badge',
  'badge--translucent',
  'badge--orange',
  'test-class',
];

export const BadgeDefaultPropsMock = {
  variant: 'translucent',
  color: 'orange',
  className: 'test-class',
};

export const BadgeChildrenMock = 'test content';
