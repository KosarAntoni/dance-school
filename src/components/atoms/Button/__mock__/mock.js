export const testFunction = jest.fn();

export const buttonDefaultClassesMock = [
  'button',
  'button--dark-green',
  'button--solid',
  'test-class',
];

export const ButtonDefaultPropsMock = {
  onClick: testFunction,
  className: 'test-class',
  variant: 'solid',
  color: 'dark-green',
  size: 20,
};

export const ButtonChildrenMock = 'test content';
