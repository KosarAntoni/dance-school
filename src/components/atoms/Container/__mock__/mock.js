export const ContainerOuterDefaultClassesMock = [
  'container__outer',
  'container--wide',
  'container--gutter',
  'outer-test-class',
];

export const ContainerDefaultClassesMock = ['container', 'test-class'];

export const ContainerDefaultPropsMock = {
  className: 'test-class',
  outerClassName: 'outer-test-class',
  variant: 'wide',
};

export const ContainerChildrenMock = 'test content';
