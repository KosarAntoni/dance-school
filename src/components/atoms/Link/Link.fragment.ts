import { graphql } from 'gatsby';

export const LinkFragment = graphql`
  fragment Link on GraphCMS_Link {
    color
    id
    hasArrow
    label
    noHover
    url
    variant
    ariaLabel
  }
`;
