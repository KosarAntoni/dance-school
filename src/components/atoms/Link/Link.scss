@use 'sass:color';
@import 'styles/component-set';

%solid {
  border-radius: 8px;
}

.link {
  text-align: left;
  text-decoration: none;
  display: inline-flex;
  padding: 0;
  background-color: transparent;
  border: none;
  transition: color $transition;

  &__arrow {
    position: relative;
    align-self: center;
    width: 1em;
    height: 1em;
    margin-left: 0.2em;
    overflow: hidden;

    @include flex-center;

    &--visible,
    &--hidden {
      font-size: 0.6em;
      position: absolute;
      transition: transform $transition, opacity $transition;
    }

    &--visible {
      opacity: 1;
      transform: translateX(0);
    }

    &--hidden {
      opacity: 0;
      transform: translateX(-100%);
    }
  }

  &:not(.link--no-hover):hover,
  &:not(.link--no-hover):focus {
    .link__arrow {
      &--visible {
        opacity: 0;
        transform: translateX(100%);
      }

      &--hidden {
        opacity: 1;
        transform: translateX(0);
      }
    }
  }

  &__content {
    display: inline-flex;
  }

  &--solid {
    &-white {
      @extend %solid;

      color: $c-white;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        background-color: color.adjust($c-white, $lightness: -15%);
      }
    }

    &-black {
      @extend %solid;

      color: $c-black;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        background-color: color.adjust($c-black, $lightness: 85%);
      }
    }

    &-blue {
      @extend %solid;

      color: $c-pastel-blue;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        background-color: color.adjust($c-pastel-blue, $lightness: 25%);
      }
    }

    &-gray {
      @extend %solid;

      color: $c-gray;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        background-color: color.adjust($c-gray, $lightness: 50%);
      }
    }
  }

  &--simple {
    &-white {
      color: $c-white;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        color: color.adjust($c-white, $lightness: -10%);
      }
    }

    &-black {
      color: $c-black;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        color: color.adjust($c-black, $lightness: 25%);
      }
    }

    &-blue {
      color: $c-pastel-blue;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        color: color.adjust($c-pastel-blue, $lightness: 5%);
      }
    }

    &-gray {
      color: $c-gray;

      &:not(.link--no-hover):hover,
      &:not(.link--no-hover):focus {
        color: color.adjust($c-gray, $lightness: 50%);
      }
    }
  }
}
