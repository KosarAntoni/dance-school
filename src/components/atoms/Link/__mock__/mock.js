export const linkDefaultClassesMock = ['link link--solid-blue', 'test-class'];

export const LinkDefaultPropsMock = {
  url: '/about',
  className: 'test-class',
  variant: 'solid',
  color: 'blue',
  size: 20,
  hasArrow: true,
};

export const LinkChildrenMock = 'test content';
