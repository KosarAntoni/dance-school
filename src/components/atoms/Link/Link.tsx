import React, { FC } from 'react';
import { HiArrowRight } from '@react-icons/all-files/hi/HiArrowRight';
import classNames from 'classnames';

import NavBase from 'atoms/NavBase';
import Typography from 'atoms/Typography';

import { LinkProps } from './models.d';

import './Link.scss';

const Link: FC<LinkProps> = ({
  children,
  color = 'black',
  variant = 'simple',
  hasArrow,
  noHover,
  className,
  onClick,
  url,
  queryString,
  ariaLabel,
  lineHeight,
  weight,
  italic,
  size,
  padding,
}) => {
  const linkClasses = classNames(
    'link',
    {
      [`link--${variant}-${color}`]: variant && color,
      'link--no-hover': url === '#' || !url || noHover,
    },
    className
  );

  const navBaseProps = { onClick, url, queryString, ariaLabel };
  const typographyProps = {
    lineHeight,
    weight,
    italic,
    size,
    padding,
    className: 'link__content',
  };

  const renderArrow = hasArrow && (
    <span className="link__arrow">
      <HiArrowRight className="link__arrow--hidden" />
      <HiArrowRight className="link__arrow--visible" />
    </span>
  );

  return (
    <NavBase className={linkClasses} {...navBaseProps}>
      <Typography as="span" {...typographyProps}>
        {children}
        {renderArrow}
      </Typography>
    </NavBase>
  );
};

export default Link;
