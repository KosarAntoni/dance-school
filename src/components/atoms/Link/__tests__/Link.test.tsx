import React from 'react';
import { Link as GatsbyLink } from 'gatsby';
import { fireEvent, render, screen } from '@testing-library/react';

import { LinkChildrenMock, linkDefaultClassesMock, LinkDefaultPropsMock } from '../__mock__/mock';
import Link from '../Link';

import { LinkProps } from '../models';

const defaultProps: LinkProps = {
  ...LinkDefaultPropsMock,
};

test('the Link component renders correctly with required and optional props', () => {
  render(<Link {...defaultProps}>{LinkChildrenMock}</Link>);

  const linkNode = screen.getByRole('link');

  // the component has been rendered
  expect(linkNode).toBeInTheDocument();

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(linkNode).toMatchSnapshot();

  // the component opens propper page after click
  fireEvent.click(linkNode);
  expect(GatsbyLink).toHaveBeenCalledWith(
    expect.objectContaining({ to: LinkDefaultPropsMock.url }),
    expect.anything()
  );

  // the comonent has correct classes
  expect(linkNode).toHaveClass(linkDefaultClassesMock.join(' '));
  expect(linkNode.firstChild).toHaveClass('typography--size-20');

  // the component has arrow
  expect(linkNode.firstChild?.lastChild).toHaveClass('link__arrow');
});
