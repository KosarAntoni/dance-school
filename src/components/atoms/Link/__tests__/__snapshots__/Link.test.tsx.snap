// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Link component renders correctly with required and optional props 1`] = `
<a
  class="link link--solid-blue test-class"
  href="/about"
>
  <span
    class="typography link__content typography--size-20 typography--line-height-normal"
    data-testid="typography-item"
  >
    test content
    <span
      class="link__arrow"
    >
      <svg
        class="link__arrow--hidden"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.293 3.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L14.586 11H3a1 1 0 110-2h11.586l-4.293-4.293a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
      <svg
        class="link__arrow--visible"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.293 3.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L14.586 11H3a1 1 0 110-2h11.586l-4.293-4.293a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </span>
</a>
`;
